### ISSUE TITLE
Build break on (at least) Mac OS X 10.12 Sierra

### ISSUE THREAD

**Issue by Birch-san** [SFTtech/openage#643]
I had a try at building this on OS X Sierra. Since I was unable to find any screenshots, video, releases or status documentation since 2015, and wanted to find out what the project looks like today.

There were 4 problems so far. I have solutions for three of them.
## Problem 1: [configure] `Qt5Core` package configuration not found

```
Call Stack (most recent call first):
  CMakeLists.txt:88 (doxygen_configure)



cpp library
    libopenage                                          [212 sources] [15 generated]
CMake Error at libopenage/CMakeLists.txt:90 (find_package):
  By not providing "FindQt5Core.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Qt5Core", but
  CMake did not find one.

  Could not find a package configuration file provided by "Qt5Core"
  (requested version 5.4) with any of the following names:

    Qt5CoreConfig.cmake
    qt5core-config.cmake

  Add the installation prefix of "Qt5Core" to CMAKE_PREFIX_PATH or set
  "Qt5Core_DIR" to a directory containing one of the above files.  If
  "Qt5Core" provides a separate development package or SDK, be sure it has
  been installed.
```

Qt5 provides in its `lib/cmake` directory (i.e. `/usr/local/Cellar/qt5/5.6.2/lib/cmake`) some `cmake` scripts (e.g. `Qt5CoreConfig.cmake`). Our `cmake` needs to find these, since they tell it how to link against Qt5's modules (e.g. `Qt5::Core`).
#### Solution

Before running `./configure`, I set the following variable in my shell:

``` bash
CMAKE_PREFIX_PATH="$CMAKE_PREFIX_PATH:/usr/local/Cellar/qt5/5.6.2/lib/cmake"
```
#### Another solution

There is an alternative solution I found that for some reason works.

That is: add `/usr/local/opt/qt5/bin` to your `$PATH`.

I think I remember that this turned out to be only a partial solution. However it's possible my memory is failing me, and that the "remaining problem" with it was simply Problem 3 (which is a problem we encounter with either of the solutions).
## Problem 2: [make] `GL/gl.h` not found

```
➜  openage git:(master) 🍔 make
…
[  1%] Building CXX object libopenage/CMakeFiles/libopenage.dir/engine.cpp.o
In file included from /Users/birch/git/openage/libopenage/engine.cpp:17:
In file included from /Users/birch/git/openage/libopenage/gui_basic.h:12:
/Users/birch/git/openage/libopenage/gui/guisys/public/gui_renderer.h:7:10: fatal error: 'GL/gl.h' file not found
#include <GL/gl.h>
         ^
1 error generated.
```

Apparently the OpenGL headers on Mac live under `OpenGL/gl.h`.
#### Solution

I helped it find the headers by changing `openage/libopenage/gui/guisys/public/gui_renderer.h`:

``` diff
+#ifndef __APPLE__
 #include <GL/gl.h>
+#else
+#include <OpenGL/gl.h>
+#endif
```
## Problem 3 [make]: `QCocoaNativeContext` not found

```
[ 34%] Building CXX object libopenage/CMakeFiles/libopenage.dir/gui/guisys/private/platforms/context_extraction_cocoa.mm.o
/Users/birch/git/openage/libopenage/gui/guisys/private/platforms/context_extraction_cocoa.mm:7:10: fatal error: 
      'QtPlatformHeaders/QCocoaNativeContext' file not found
#include <QtPlatformHeaders/QCocoaNativeContext>
         ^
1 error generated.
```

So, it needs more includes. I used `make VERBOSE=1` to check what includes we were already recruiting:

``` bash
[ 34%] Building CXX object libopenage/CMakeFiles/libopenage.dir/gui/guisys/private/platforms/context_extraction_cocoa.mm.o
cd /Users/birch/git/openage/.bin/clang++-3-5-debug-Oauto-sanitize-none/libopenage && /usr/local/bin/clang++-3.5   -DQT_CORE_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_QML_LIB -DQT_QUICK_LIB -Dlibopenage_EXPORTS -I/Library/Frameworks/Mono.framework/Headers -I/usr/local/include/freetype2 -I/usr/local/include -I/usr/local/include/opus -I/Users/birch/Library/Frameworks/SDL2.framework/Headers -I/Library/Frameworks/SDL2_image.framework/Headers -iframework /usr/local/Cellar/qt5/5.6.2/lib -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtCore.framework/Headers -isystem /usr/local/Cellar/qt5/5.6.2/./mkspecs/macx-clang -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtQuick.framework/Headers -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtGui.framework/Headers -isystem /System/Library/Frameworks/OpenGL.framework/Headers -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtQml.framework/Headers -isystem /usr/local/Cellar/qt5/5.6.2/lib/QtNetwork.framework/Headers  -Wall -Wextra -pedantic -Wno-mismatched-tags -stdlib=libc++ -std=c++14 -g -O1 -fPIC   -F/Users/birch/Library/Frameworks  -fPIC -std=gnu++14 -o CMakeFiles/libopenage.dir/gui/guisys/private/platforms/context_extraction_cocoa.mm.o -c /Users/birch/git/openage/libopenage/gui/guisys/private/platforms/context_extraction_cocoa.mm
```

So, we already have quite a few Qt header directories included. Such as the `Qt5::Core` module's headers, in the directory: `/usr/local/Cellar/qt5/5.6.2/lib/QtCore.framework/Headers`.

I noticed that Qt provide _no_ `QtPlatformHeaders.framework` module. So we will have to use a different strategy here.
#### Solution:

I added to the `libopenage`'s `CMakeLists.txt`

``` diff
+if(APPLE)
+   set(QTPLATFORM_INCLUDE_DIRS "/usr/local/Cellar/qt5/5.6.2/include")
+endif()

# directories for header inclusion
include_directories(
    ${OPENGL_INCLUDE_DIR}
    ${FREETYPE_INCLUDE_DIRS}
    ${EPOXY_INCLUDE_DIRS}
    ${OPUS_INCLUDE_DIRS}
    ${SDL2_INCLUDE_DIR}
    ${SDL2IMAGE_INCLUDE_DIRS}
    ${HarfBuzz_INCLUDE_DIRS}
+   ${QTPLATFORM_INCLUDE_DIRS}
)
```

Sadly this is hardcoding. Maybe there is a better way. It certainly had the desired effect: `clang` was able to find the `QtPlatformHeaders/QCocoaNativeContext` headers.
### Problem 4 [make]: clang gets very confused trying to read OS X `Foundation`'s `NSObject` header

I am not sure what's going on here.

```
[ 34%] Building CXX object libopenage/CMakeFiles/libopenage.dir/gui/guisys/private/platforms/context_extraction_cocoa.mm.o
In file included from /Users/birch/git/openage/libopenage/gui/guisys/private/platforms/context_extraction_cocoa.mm:7:
In file included from /usr/local/Cellar/qt5/5.6.2/include/QtPlatformHeaders/QCocoaNativeContext:1:
In file included from /usr/local/Cellar/qt5/5.6.2/include/QtPlatformHeaders/qcocoanativecontext.h:38:
In file included from /System/Library/Frameworks/AppKit.framework/Headers/NSOpenGL.h:9:
In file included from /System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:10:
In file included from /System/Library/Frameworks/Foundation.framework/Headers/NSArray.h:5:
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:19:21: error: expected a type
- (id)copyWithZone:(nullable NSZone *)zone;
                    ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:25:28: error: expected a type
- (id)mutableCopyWithZone:(nullable NSZone *)zone;
                           ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:32:4: error: expected a type
- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder; // NS_DESIGNATED_INITIALIZER
   ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:44:12: error: unknown property attribute 'class'
@property (class, readonly) BOOL supportsSecureCoding;
           ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:57:4: error: expected a type
- (nullable id)replacementObjectForCoder:(NSCoder *)aCoder;
   ^
/System/Library/Frameworks/Foundation.framework/Headers/NSObject.h:58:4: error: expected a type
- (nullable id)awakeAfterUsingCoder:(NSCoder *)aDecoder NS_REPLACES_RECEIVER;
   ^
In file included from /Users/birch/git/openage/libopenage/gui/guisys/private/platforms/context_extraction_cocoa.mm:7:
In file included from /usr/local/Cellar/qt5/5.6.2/include/QtPlatformHeaders/QCocoaNativeContext:1:
In file included from /usr/local/Cellar/qt5/5.6.2/include/QtPlatformHeaders/qcocoanativecontext.h:38:
In file included from /System/Library/Frameworks/AppKit.framework/Headers/NSOpenGL.h:9:
In file included from /System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:10:
In file included from /System/Library/Frameworks/Foundation.framework/Headers/NSArray.h:6:
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:7:15: error: expected ';' after @class
@class NSArray<ObjectType>;
              ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:7:16: error: cannot find protocol declaration for 'ObjectType'
@class NSArray<ObjectType>;
               ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:34:25: error: cannot find protocol declaration for 'ObjectType'
@interface NSEnumerator<ObjectType> : NSObject <NSFastEnumeration>
                        ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:34:37: error: expected unqualified-id
@interface NSEnumerator<ObjectType> : NSObject <NSFastEnumeration>
                                    ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:36:4: error: expected a type
- (nullable ObjectType)nextObject;
   ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:40:25: error: cannot find protocol declaration for 'ObjectType'
@interface NSEnumerator<ObjectType> (NSExtendedEnumerator)
                        ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:40:1: error: duplicate interface definition for class 'NSEnumerator'
@interface NSEnumerator<ObjectType> (NSExtendedEnumerator)
^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:34:12: note: previous definition is here
@interface NSEnumerator<ObjectType> : NSObject <NSFastEnumeration>
           ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:40:37: error: method type specifier must start with '-' or '+'
@interface NSEnumerator<ObjectType> (NSExtendedEnumerator)
                                    ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:40:38: error: expected a type
@interface NSEnumerator<ObjectType> (NSExtendedEnumerator)
                                     ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:42:1: error: expected selector for Objective-C method
@property (readonly, copy) NSArray<ObjectType> *allObjects;
^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:42:28: error: unknown type name 'NSArray'
@property (readonly, copy) NSArray<ObjectType> *allObjects;
                           ^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:42:35: error: expected member name or ';' after declaration specifiers
@property (readonly, copy) NSArray<ObjectType> *allObjects;
                           ~~~~~~~^
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:42:1: error: property requires fields to be named
@property (readonly, copy) NSArray<ObjectType> *allObjects;
^                          ~~~~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

I think it doesn't understand the syntax of these headers (Foundation is a system framework that comes with OS X). So maybe it was expecting C++, but encountered Objective-C, and exploded? But at the same time: it is aware of Objective-C, as evidenced by the message:

> expected selector for Objective-C method

So, I don't really get what happened. Anyway, I thought I may as well report my findings.

**Comment by ChipmunkV** [comment]
Yes, we really need a Mac OS X developer to fix the Qt integration. I was over-confident while implementing Mac support without Mac.

**Comment by TheJJ** [comment]
Is it possible to use C++ instead of Objective-C for the context extraction if you suspect language conflicts?

**Comment by Birch-san** [comment]
> Is it possible to use C++ instead of Objective-C for the context extraction if you suspect language conflicts?
Maybe? I've never seen how to do something like that.

**Comment by Birch-san** [comment]
```
/System/Library/Frameworks/Foundation.framework/Headers/NSEnumerator.h:7:15: error: expected ';' after @class
@class NSArray<ObjectType>;
```

That particular error suggests to me that the compiler has never heard of generic types. So certainly I suspect it's encountering an unexpected programming language.

**Comment by Longhanks** [comment]
Extracting the context still seems like a very, very tacky workaround. X11 is currently the only working platform, Macs fail (as seen here) and I highly doubt this will work with Wayland, although I haven't tested it.

Is this really the way to go forward?

**Comment by ChipmunkV** [comment]
> X11 is currently the only working platform
windows should be fine (if not for the python situation)

> Is this really the way to go forward?

Both Qt and SDL are cross-platform libraries. They know everything about every platform, but don't know anything about each-other.

To use a gl context or a window id, the handle has to be converted from an SDL wrapper to its platform-specific representation and then wrapped into a Qt wrapper.

> Extracting the context still seems like a very, very tacky workaround

Another workaround was to create a separate context. But when you've tested that on Mac, there was a problem with the FBO initialisation (which is odd because it's a totally usual use case).

**Comment by shobhitg** [comment]
> Both Qt and SDL are cross-platform libraries. They know everything about every platform, but don't know anything about each-other.

> To use a gl context or a window id, the handle has to be converted from an SDL wrapper to its platform-specific representation and then wrapped into a Qt wrapper.

I thought that qt supported OpenGL without the need of any SDL wrapper.

I know that SDL is framebuffer lib. What is SDL used for in this project?

**Comment by ChipmunkV** [comment]
We are opposed to the frameworks, so everything is used as a library: SDL for OpenGL initialization and input, Qt - for GUI.

**Comment by shobhitg** [comment]
@ChipmunkV 
I just wanted to have a feel of it, can you point me to the folder that mostly contains the SDL based opengl work?

**Comment by ChipmunkV** [comment]
Window and GL context creation, input/game loop:
https://github.com/SFTtech/openage/blob/master/libopenage/engine.cpp

**Comment by Longhanks** [comment]
Which would be entirely possible with Qt. SDL really feels redundant when using Qt, in my opinion. (http://doc.qt.io/qt-5/qopenglwidget.html#details)

StackOverflow also recommends not using both frameworks/libraries together:
http://stackoverflow.com/a/27345513

**Comment by TheJJ** [comment]
We really want to avoid Qt stuff spreading all over the codebase. If there's another update to Qt 6 or whatever, it will be more painfull the more we tightly integrated that stuff.

Also, we want to handle the loop ourselves and not pass control to Qt, which is impossible when using a QWindow or whatever it is named and embedding the OpenGL widget. Furthermore, by not depending on Qt for drawing the actual game, we can provide other drawing backends (html5, vulkan, ...) more easily (how we paint the GUI then is another thing, but you get the point.

Stackoverflow also doesn't "recommend" not doing it, it's just a post of datenwolf (btw i know him personally :). And it was back in 2014.

**Comment by Longhanks** [comment]
More painful than creating a "context extraction" for every platform that may or may not break depending on internal APIs? I mean, the current context extraction already fails on OS X and will probably not work on Wayland or Mir. Qt already has a Wayland backend, for example.

It's a matter of trade offs, I understand that. But Qt abstracting away the underlaying platform APIs seems worth granting it the tradeoff, I'd say - is there a reason you definitely need 100% control of the run loop? Or is it impossible to implement the necessary calls via the Qt APIs? (Genuinely asking :) )

**Comment by TheJJ** [comment]
Probably you're right, we should look into this if it makes the integration easier. On the other hand, then the path to other renderers is probably blocked once we create a Qt Window with the QtOpenGLContext. But I'm no expert in that matter :)

**Comment by Birch-san** [comment]
Looks like the problem was that I was using qt 5.6. Not sure how that happened; maybe I skipped the `brew update` step and ended up with out-of-date package definitions when I began.

Now that I have `brew upgrade qt5`d to qt 5.7.0: my machine now successfully builds. Instead of using any of the code changes and env variables described in this issue, I tried a build using dusek's branch (#658). It succeeded using the standard steps.
